### PR TITLE
Fixes 96: Add bulk create

### PR DESCRIPTION
**Pull request by rverdile**
Adds bulk create

**Example:**
POST `/api/content_sources/v1/repositories/bulk_create`

Body:
```
[
  {
    "name": "test1",
    "url": "test1",
  },
  {
    "name": "test2",
    "url": "test2",
  }
]
```

Response:
```
[
    {
       "error": nil,
       "repository": {...repo}
    },
    
    {
    
       "error":nil,
       "repository": {...repo}
    }, ...
 ]

OR when there's an error

[
  {
       "error": "error",
       "repository": nil,
    },
{
       "error": nil,
       "repository": nil,
    },

]
```

Still needs

- [x] Handler tests
- [x] Fix error response

### CODE REVIEW

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMSCONTENT-96

**Comment by jlsherrill** [comment]
The changes I made to the tests to delete the data from the 3 tables in my
rh repo load pr will probably solve the transaction errors.  The order of
deletion was wrong

**Comment by rverdile** [comment]
I fixed that here too. I think it has something to do with there being a transaction both in the test and in the dao layer.

**Comment by jlsherrill** [comment]
ahh, maybe try nested transactions? https://gorm.io/docs/transactions.html#Nested-Transactions

File: pkg/handler/repositories.go [hunk 1]
```diff
@@ -95,6 +99,53 @@ func (rh *RepositoryHandler) createRepository(c echo.Context) error {
...
+	}
+
+	var response []api.RepositoryResponse
+
+	if err = db.DB.Transaction(func(tx *gorm.DB) error {
+		response, err = rh.RepositoryDao.BulkCreate(tx, newRepositories)
+		return err
+	}); err != nil {
+		return echo.NewHTTPError(httpCodeForError(err), err.Error())
+	}
```
[the remark this answers is not in the document]
Comment (jlsherrill) [reply pkg/handler/repositories.go:144]: This seemed to work for me:  https://github.com/rverdile/content-sources-backend/pull/1   
(at least it didn't cause any tests to fail)

(opened a PR on your branch for easier viewing)

File: pkg/handler/repositories.go [hunk 2]
```diff
@@ -95,6 +99,53 @@ func (rh *RepositoryHandler) createRepository(c echo.Context) error {
...
 }
 
+// CreateRepository godoc
+// @Summary      Bulk create repositories
+// @ID           bulkCreateRepositories
+// @Description  bulk create repositories
+// @Tags         repositories
+// @Accept       json
+// @Produce      json
+// @Param        body  body     api.RepositoryRequest  true  "request body"
```
Comment (jlsherrill) [inline pkg/handler/repositories.go:109]: This should be an array of api.RepositoryRequest right?

File: pkg/handler/repositories.go [hunk 3]
```diff
@@ -95,6 +99,53 @@ func (rh *RepositoryHandler) createRepository(c echo.Context) error {
...
+// @Success      201  {object}  api.RepositoryResponse
+// @Header       201  {string}  Location "resource URL"
+// @Router       /repositories/ [post]
```
Comment (jlsherrill) [inline pkg/handler/repositories.go:112]: /repositories/bulk_create/

**Comment by jlsherrill** [comment]
One thing to note, is the card describes a response that looks like this:
{ "RepoName": {"error": "Error Found"}, "2nd repo name": {"error":nil}}

but the current pr doesn't follow that.  You might wanna check with @Andrewgdewar as he and i discussed this previously. 

We could also return the repo that was created as part of it within that structure:

{ "RepoName": {"error": "Error encountered", "repository": nil}, 
"2nd repo name": {"error":nil, "repository": {"uuid":"abcd"...}}}

**Comment by rverdile** [comment]
Also, should it be an array? 
```
[
  {  "repoName1": 
   {
       "error": "error1",
        "repo": nil
    }
  },
  {  "repoName2": 
   {
       "error": nil,
        "repo": {repo...}
    }
  },
]
```

And to clarify, this would also return the successful repos on error? Not just the ones that errored?

**Comment by jlsherrill** [comment]
repo names are supposed to be unique, so i think just  just one big hash where the name is the key is fine, you may want to preemptively make sure that no duplicates are passed in though and reject the request outright.  IDK @Andrewgdewar your thoughts?

**Comment by jlsherrill** [comment]
Built and pushed image for 5f7208ad3d802c4c0fd34681b31421766e388de4

**Comment by rverdile** [comment]
ignore the push this doesn't work yet

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/14702687

**Comment by jlsherrill** [comment]
Built and pushed image for 7da566d8d8890af896b360f3324501eda3e47610

**Comment by rverdile** [comment]
should be all updated now

**Comment by jlsherrill** [comment]
Built and pushed image for 0c4064ad456845d0a94b94c5ee732a7357ca3fc4

**Comment by jlsherrill** [comment]
Built and pushed image for 910fca4a69f1b209e4991d9ec263cab7d437d295

File: pkg/handler/repositories.go [hunk 4]
```diff
@@ -11,6 +12,8 @@ import (
 	"github.com/redhatinsights/platform-go-middlewares/identity"
 )
 
+const BulkCreateLimit = 20
```
Comment (jlsherrill) [inline pkg/handler/repositories.go:15]: i created 20 repos and it took ~19 ms.  💨

**Review by jlsherrill — APPROVED** [review]

File: pkg/handler/repositories.go [hunk 5]
```diff
@@ -95,6 +99,47 @@ func (rh *RepositoryHandler) createRepository(c echo.Context) error {
...
+// @Summary      Bulk create repositories
+// @ID           bulkCreateRepositories
+// @Description  bulk create repositories
+// @Tags         repositories
+// @Accept       json
+// @Produce      json
+// @Param        body  body     []api.RepositoryRequest  true  "request body"
+// @Success      201  {object}  []api.RepositoryBulkCreateResponse
+// @Header       201  {string}  Location "resource URL"
+// @Router       /repositories/bulk_create/ [post]
```
Comment (rverdile) [inline pkg/handler/repositories.go:112]: updated openapi

**Comment by jlsherrill** [comment]
Built and pushed image for 0b2b0940b480543cda033bb27cd468d521fd0913

**Review by swadeley — APPROVED** [review]
ACK

**Pull request merged**